edk2toollib/uefi/edk2/path_utilities.py

```diff
@@ -19,64 +19,91 @@
 
 
 class Edk2Path(object):
+    """Represents edk2 file paths.
 
-    def __init__(self, ws: os.PathLike, packagepathlist: Iterable[os.PathLike], error_on_invalid_pp: bool = True):
-        """ An Edk2Path object is an object that can be used to resolve edk2 relative paths
+    Can be used to resolve relative paths.
+    """
+
+    def __init__(self, ws: os.PathLike, package_path_list: Iterable[os.PathLike], error_on_invalid_pp: bool = True):
+        """Constructor.
 
         Args:
-            ws: absolute path or cwd relative path of the workspace.
-            packagespathlist: list of packages path.
+            ws (os.PathLike): absolute path or cwd relative path of the workspace.
+            packagespathlist (Iterable[os.PathLike]): list of packages path.
                 Entries can be Absolute path, workspace relative path, or CWD relative.
-            error_on_invalid_pp: default value is True. If packages path value is invalid raise exception
+            error_on_invalid_pp (bool): default value is True. If packages path value is invalid raise exception.
+
+        Raises:
+            (NotADirectoryError): Invalid workspace or package path directory.
         """
 
-        self.WorkspacePath = ws
         self.logger = logging.getLogger("Edk2Path")
-        if (not os.path.isabs(ws)):
-            self.WorkspacePath = os.path.abspath(os.path.join(os.getcwd(), ws))
-
-        if (not os.path.isdir(self.WorkspacePath)):
-            self.logger.error("Workspace path invalid.  {0}".format(ws))
-            raise Exception("Workspace path invalid.  {0}".format(ws))
-
-        # Set PackagePath
-        self.PackagePathList = list()
-        for a in packagepathlist:
-            if (os.path.isabs(a)):
-                self.PackagePathList.append(a)
+
+        # Other code is dependent the following types, so keep it that way:
+        #   - self.PackagePathList: List[str]
+        #   - self.WorkspacePath: str
+
+        self.PackagePathList = []
+        self.WorkspacePath = ""
+
+        workspace_candidate_path = Path(ws)
+
+        if not workspace_candidate_path.is_absolute():
+            workspace_candidate_path = Path(os.getcwd(), ws)
+
+        if not workspace_candidate_path.is_dir():
+            raise NotADirectoryError(
+                errno.ENOENT,
+                os.strerror(errno.ENOENT),
+                workspace_candidate_path.resolve())
+
+        self.WorkspacePath = str(workspace_candidate_path)
+
+        candidate_package_path_list = []
+        for a in package_path_list:
+            if os.path.isabs(a):
+                candidate_package_path_list.append(Path(a))
             else:
-                # see if workspace relative
-                wsr = os.path.join(ws, a)
-                if (os.path.isdir(wsr)):
-                    self.PackagePathList.append(wsr)
+                wsr = Path(ws, a)
```

Comment:
shouldn't this use self.WorkspacePath instead of ws given that ws could be cwd relative.

Reply:
Yeah, that's better. Updated in PR commit https://github.com/tianocore/edk2-pytool-library/pull/186/commits/e910271effdf999b4c770746f8ebd52cbd3fb0fb